examples/src/socketio-retry/socketio-retry.rs

```diff
@@ -0,0 +1,82 @@
+//! This a end to end test server used with this [test suite](https://github.com/socketio/socket.io-protocol)
+
+use std::time::Duration;
+
+use hyper::Server;
+use serde_json::Value;
+use socketioxide::errors::{AckSenderError, BroadcastError, SendError, TransportError};
+use socketioxide::{Namespace, SocketIoConfig, SocketIoService};
+use tokio::time::sleep;
+use tracing::{info, Level};
+use tracing_subscriber::FmtSubscriber;
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let subscriber = FmtSubscriber::builder()
+        .with_line_number(true)
+        .with_max_level(Level::INFO)
+        .finish();
+    tracing::subscriber::set_global_default(subscriber)?;
+
+    let config = SocketIoConfig::builder()
+        .ping_interval(Duration::from_millis(300))
+        .ping_timeout(Duration::from_millis(200))
+        .max_payload(1e6 as u64)
+        // set buffer size, it will force TransportError
+        .max_buffer_size(1usize)
+        .build();
+    info!("Starting server");
+
+    let ns = Namespace::builder()
+        .add("/", |socket| async move {
+            info!("Socket.IO connected: {:?} {:?}", socket.ns(), socket.sid);
+            let data: Value = socket.handshake.data().unwrap();
+            socket.emit("auth", data).ok();
+
+            socket.on("message", |socket, data: Value, bin, _| async move {
+                info!("Received event: {:?} {:?}", data, bin);
+                if let Err(BroadcastError::SendError(mut errors)) =
+                    socket.bin(bin).emit("message-back", data)
+                {
+                    let err = errors.pop().unwrap();
+                    if let SendError::TransportError(TransportError::SendFailedBinPayloads(_)) = err
+                    {
+                        while let Err(TransportError::SendFailedBinPayloads(_)) =
+                            socket.retry_failed()
+                        {
+                            sleep(Duration::from_millis(10)).await;
+                            info!("retry");
+                        }
+                    }
+                }
+            });
+
+            socket.on("message-with-ack", |_, data: Value, bin, ack| async move {
+                info!("Received event: {:?} {:?}", data, bin);
+                if let Err(AckSenderError::SendError {
+                    send_error: SendError::TransportError(TransportError::SendFailedBinPayloads(_)),
+                    socket,
+                }) = ack.bin(bin).send(data)
+                {
+                    while let Err(TransportError::SendFailedBinPayloads(_)) = socket.retry_failed()
+                    {
+                        sleep(Duration::from_millis(10)).await;
+                        info!("retry");
+                    }
+                }
+            });
```

Comment:
The control flow is hard and the retry mecanism complex... It requires some comments so the reader easily understand what is happening